solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformer.java

```diff
@@ -87,7 +87,12 @@ public void transform(SolrDocument rootDoc, int rootDocId) {
       final int segBaseId = leafReaderContext.docBase;
       final int segRootId = rootDocId - segBaseId;
       final BitSet segParentsBitSet = parentsFilter.getBitSet(leafReaderContext);
-      final int segPrevRootId = segParentsBitSet.prevSetBit(segRootId - 1); // can return -1 and that's okay
+      final int segPrevRootId = rootDocId==0? -1: segParentsBitSet.prevSetBit(segRootId - 1); // can return -1 and that's okay
+
+      if(segPrevRootId == (rootDocId - 1)) {
```

Comment:
you are comparing a segment local ID with a global ID which is incorrect.  You should refer to segRootId.  This is why I'm particular about using "seg" nomenclature in a body of code that deals with both segment and global IDs -- it makes it at least easier to identify such an error.  It's difficult to get tests to detect this; we'd need to commit some docs up front to cause more segments to be created than many tests will do.

Reply:
You altered line 90 in response to my comment.  I'm referring to line 92 -- `if(segPrevRootId == (rootDocId - 1))`, where my comment is.

Interesting though.... line 90 is different from the line of code I committed to the feature branch.  That line on the feature branch is:
`final int segPrevRootId = segParentsBitSet.prevSetBit(segRootId - 1); // can return -1 and that's okay`
Notice there is no conditional, but there is in your version in this PR.  BitSet.prevSetBit will return -1 if given a 0 input (so says it's documentation).  That's what I was trying to say in my comment on the line of code.  Why did you change it?

Reply:
If segRootId is 0, segParentsBitSet.prevSetBit(-1) throws an assertion error, since the index has to be >= 0.
I will fix line 92

Reply:
Ooooh, good catch.

Lets enhance the tests in this file a bit to help us give confidence that we're using docIDs correctly (and help avoid future enhancers/modifiers from introducing similar bugs).  Here's what I propose:  in the @BeforeClass, if random().nextBoolean(), add some nested docs -- using one of your existing nested document adding methods.  And randomly do a commit() to flush the segment.  Later in the test methods we need to add a filter query that will exclude those docs.  One way to do this is to ensure these first docs have some field we can exclude.  Another way might be knowing the maximum uniqueKey ID you can query by prior to the test starting, and then adding a filter query with a range saying the uniqueKey must be at least this value.  Make sense?

Reply:
Yes,
I have added this conditional to the test.
Let's hope I did not misread your intentions :).